enterprise/server/raft/filestore/filestore.go

```diff
@@ -35,7 +34,111 @@ type WriteCloserMetadata interface {
 	MetadataWriter
 }
 
-func NewWriter(ctx context.Context, fileDir string, wb pebble.Writer, fileRecord *rfpb.FileRecord) (WriteCloserMetadata, error) {
+// returns partitionID, isolation, hash or
+// returns partitionID, isolation, remote_instance_name, hahs
+func fileRecordSegments(r *rfpb.FileRecord) (partID string, groupID string, isolation string, remoteInstanceHash string, digestHash string, err error) {
+	if r.GetIsolation().GetPartitionId() == "" {
+		err = status.FailedPreconditionError("Empty partition ID not allowed in filerecord.")
+		return
+	}
+	partID = r.GetIsolation().GetPartitionId()
+	groupID = r.GetIsolation().GetGroupId()
+
+	if r.GetIsolation().GetCacheType() == rfpb.Isolation_CAS_CACHE {
+		isolation = "cas"
+	} else if r.GetIsolation().GetCacheType() == rfpb.Isolation_ACTION_CACHE {
+		isolation = "ac"
+		if remoteInstanceName := r.GetIsolation().GetRemoteInstanceName(); remoteInstanceName != "" {
+			remoteInstanceHash = strconv.Itoa(int(crc32.ChecksumIEEE([]byte(remoteInstanceName))))
+		}
+	} else {
+		err = status.FailedPreconditionError("Isolation type must be explicitly set, not UNKNOWN.")
+		return
+	}
+	if len(r.GetDigest().GetHash()) <= 4 {
+		err = status.FailedPreconditionError("Malformed digest; too short.")
+		return
+	}
+	digestHash = r.GetDigest().GetHash()
+	return
+}
+
+type Store interface {
+	FileKey(r *rfpb.FileRecord) ([]byte, error)
+	FilePath(fileDir string, f *rfpb.StorageMetadata_FileMetadata) string
+	FileMetadataKey(r *rfpb.FileRecord) ([]byte, error)
+
+	NewReader(ctx context.Context, fileDir string, md *rfpb.StorageMetadata, offset, limit int64) (io.ReadCloser, error)
+	NewWriter(ctx context.Context, fileDir string, wb pebble.Writer, fileRecord *rfpb.FileRecord) (WriteCloserMetadata, error)
+
+	InlineReader(f *rfpb.StorageMetadata_InlineMetadata, offset, limit int64) (io.ReadCloser, error)
+	InlineWriter(ctx context.Context, sizeBytes int64) WriteCloserMetadata
+
+	FileReader(ctx context.Context, fileDir string, f *rfpb.StorageMetadata_FileMetadata, offset, limit int64) (io.ReadCloser, error)
+	FileWriter(ctx context.Context, fileDir string, fileRecord *rfpb.FileRecord) (WriteCloserMetadata, error)
+}
+
+type fileStorer struct {
+	includeGroupIDInFilePaths bool
+}
+
+func New(includeGroupIDInFilePaths bool) Store {
```

Comment:
What do you think about calling it just `useGroupIDs` or `isolateByGroupIDs` as this opt affects keys as well?

Reply:
Done.